crates/topos-p2p/src/behaviour/gossip.rs

```diff
@@ -0,0 +1,282 @@
+use std::{
+    collections::{HashSet, VecDeque},
+    env,
+    task::Poll,
+    time::Duration,
+};
+
+use libp2p::{
+    gossipsub::{self, IdentTopic, Message, MessageAuthenticity, MessageId, Topic},
+    identity::Keypair,
+    swarm::{NetworkBehaviour, THandlerInEvent, ToSwarm},
+};
+use serde::{Deserialize, Serialize};
+use topos_metrics::{P2P_DUPLICATE_MESSAGE_ID_RECEIVED, P2P_GOSSIP_BATCH_SIZE};
+
+use crate::{constant, event::ComposedEvent, TOPOS_ECHO, TOPOS_GOSSIP, TOPOS_READY};
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub(crate) struct Batch {
+    pub(crate) data: Vec<Vec<u8>>,
+}
+
+pub struct Behaviour {
+    batch_size: usize,
+    gossipsub: gossipsub::Behaviour,
+    echo_queue: VecDeque<Vec<u8>>,
+    ready_queue: VecDeque<Vec<u8>>,
+    tick: tokio::time::Interval,
+    cache: HashSet<MessageId>,
+}
+
+impl Behaviour {
+    pub fn publish(&mut self, topic: &'static str, data: Vec<u8>) -> Result<usize, &'static str> {
+        match topic {
+            TOPOS_GOSSIP => {
+                _ = self.gossipsub.publish(IdentTopic::new(topic), data);
+            }
+            TOPOS_ECHO => self.echo_queue.push_back(data),
+            TOPOS_READY => self.ready_queue.push_back(data),
+            _ => return Err("Invalid topic"),
+        }
+
+        Ok(0)
+    }
+
+    pub fn subscribe(&mut self) -> Result<(), &'static str> {
+        self.gossipsub
+            .subscribe(&gossipsub::IdentTopic::new(TOPOS_GOSSIP))
+            .unwrap();
+
+        self.gossipsub
+            .subscribe(&gossipsub::IdentTopic::new(TOPOS_ECHO))
+            .unwrap();
+
+        self.gossipsub
+            .subscribe(&gossipsub::IdentTopic::new(TOPOS_READY))
+            .unwrap();
+
+        Ok(())
+    }
+
+    pub fn new(peer_key: Keypair) -> Self {
+        let batch_size = env::var("TOPOS_GOSSIP_BATCH_SIZE")
+            .map(|v| v.parse::<usize>())
+            .unwrap_or(Ok(10))
+            .unwrap();
+        let gossipsub = gossipsub::ConfigBuilder::default()
+            .max_transmit_size(2 * 1024 * 1024)
+            .validation_mode(gossipsub::ValidationMode::Strict)
+            .build()
+            .unwrap();
+
+        let gossipsub = gossipsub::Behaviour::new_with_metrics(
+            MessageAuthenticity::Signed(peer_key.clone()),
+            gossipsub,
+            constant::METRIC_REGISTRY
+                .try_lock()
+                .unwrap()
```

Comment:
Should we use `unwrap` here (and on similar places) or some friendlier error handling?

Reply:
In this place we can has it is scoped to a single used metric dedicated only for that call. We can `expect` for a better error message tho